### ISSUE TITLE
expression failed after migration from 12.1.0

### ISSUE THREAD

**Issue by shmsula** [josdejong/mathjs#3143]
**Describe the bug**
Next code stop working after migration from 12.1.0 to 12.2.0 or 12.3.0

f1(x) = x < 2 or x > 3
filter([1,2,3,4,5], f1)

**To Reproduce**
Can be simply reproduce on main [mathjs](https://mathjs.org/) page in demo widget. (math.version 12.3.0)

![image](https://github.com/josdejong/mathjs/assets/133093698/d383e7b3-9879-4fd1-8f31-daed98b9d24d)

however it works if remove "or"

![image](https://github.com/josdejong/mathjs/assets/133093698/95f7b995-e813-4ce8-b4e3-a2d94f17a34f)

On https://mathnotepad.com/ (math.version 10.3.0) works.
![image](https://github.com/josdejong/mathjs/assets/133093698/174ade72-4207-4db2-8ecb-c4c0efa3bb37)

**Comment by dvd101x** [comment]
It seems related to declaring a function in the parser that includes `or` or `and` as syntax.

```jl
X = [1, 2, 3, 4, 5]
f1(x) = x < 2 or x > 3
filter(X, f1)              # TypeError: Function filter cannot apply callback arguments to function f1: Undefined symbol x
f1(1)                      # Error: Undefined symbol x

# gets fixed if the function includes `or` as a function (not as part of the syntax)
f2(x) = or(x < 2, x > 3)
filter(X, f2)              # [1, 4, 5]

# also fixed if the function is stated inline
filter(X, x < 2 or x > 3)  # [1, 4, 5]

# also works by indexing with an array of booleans
X[X < 2 or X > 3]          # [1, 4, 5]]
```

It might have something to do with #3090

**Comment by shmsula** [comment]
The problem is that brackets aren’t required for such simple expressions, and the workaround isn’t really obvious.  I have a feeling that some kind of breaken changes were introduced, perhaps accidentally, by a commit from the 12.2.0 release. https://github.com/josdejong/mathjs/compare/v12.1.0...v12.2.0  Even a commit with a suspiciously similar description exists.

**Comment by dvd101x** [comment]
I agree it seems like a breaking change and probably unintended. I just tested what still works and what doesn't to help in the debugging process.

**Comment by josdejong** [comment]
Thanks for reporting. I think indeed that this is an unintended bug introduced in `v12.2.0` via #3090. It looks as if the `scope` isn't passed to the new transform functions for the operators `or`, `and`, `bitOr`, and `bitAnd`, resulting in this "Undefined symbol X" error.

**Comment by josdejong** [comment]
I did a bit of debugging and I do understand the cause:

In `FunctionNode`, we create a subscope when dealing with a `rawArgs` function, so the function variable `x` will not pollute the main scope and only exists in the subscope, see [FunctionNode.js#L172](https://github.com/josdejong/mathjs/blob/develop/src/expression/node/FunctionNode.js#L172).

In `OperatorNode`, we now also have to deal with rawArgs functions now (since the introduction of lazy evaluation of `and` and `or`), see [OperatorNode.js#L312](https://github.com/josdejong/mathjs/blob/develop/src/expression/node/OperatorNode.js#L312). Right now, we do not create a subscope there, since we are not (always) inside a function scope, like in the expression `(a=false) and (b=true)`. If we do create a subscope in OperatorNode, the variables `a` and `b` will be created in this subscope instead of the main scope.

I still have to think through how to solve this 🤔 . Help would be welcome.

**Comment by josdejong** [comment]
I created a branch [fix/issue3143](https://github.com/josdejong/mathjs/tree/fix/issue3143) that has unit tests for this (not yet a solution), see https://github.com/josdejong/mathjs/compare/develop...fix/issue3143

**Comment by josdejong** [comment]
Solution is almost there (just need to work out unit tests and do some refactoring and performance tests)
